File: python/vd/tracker.py
"""Centroid tracker with stable IDs and speed estimation.

Fixes the original's index-based IDs (which reshuffled every frame). Associates
detections across frames by nearest centroid with a distance gate, assigns stable
IDs, and estimates speed from centroid displacement with a pixel->metre
calibration and EMA smoothing.
"""

import math


def _centroid(box):
    x1, y1, x2, y2 = box
    return ((x1 + x2) / 2.0, (y1 + y2) / 2.0)


def _dist(a, b):
    return math.hypot(a[0] - b[0], a[1] - b[1])


class Track:
    __slots__ = ("id", "box", "cls", "score", "centroid", "last_time",
                 "speed_kmh", "max_speed", "missed", "hits",
                 "color", "color_hex", "plate", "plate_box", "thumb", "plate_thumb")

    def __init__(self, tid, box, cls, score, t):
        self.id = tid
        self.box = box
        self.cls = cls
        self.score = score
        self.centroid = _centroid(box)
        self.last_time = t
        self.speed_kmh = 0.0
        self.max_speed = 0.0
        self.missed = 0
        self.hits = 1
        self.color = None
        self.color_hex = None
        self.plate = None
        self.plate_box = None
        self.thumb = None
        self.plate_thumb = None


class Tracker:
    def __init__(self, meters_per_pixel=0.05, max_distance=120,
                 max_missed=12, smoothing=0.4):
        self.meters_per_pixel = meters_per_pixel
        self.max_distance = max_distance
        self.max_missed = max_missed
        self.smoothing = smoothing
        self.tracks = {}
        self._next_id = 1

    def set_calibration(self, mpp):
        self.meters_per_pixel = mpp

    def update(self, detections, t):
        """detections: list of (box, cls, score); t: timestamp in seconds.

        Returns list of active Track objects (matched this frame).
        """
        ids = list(self.tracks.keys())
        unmatched = set(range(len(detections)))

        pairs = []
        for tid in ids:
            tr = self.tracks[tid]
            best_i, best_d = -1, float("inf")
            for i in unmatched:
                d = _dist(tr.centroid, _centroid(detections[i][0]))
                if d < best_d:
                    best_d, best_i = d, i
            if best_i != -1 and best_d <= self.max_distance:
                pairs.append((best_d, tid, best_i))
        pairs.sort(key=lambda p: p[0])

        used_t, used_d = set(), set()
        for _, tid, i in pairs:
            if tid in used_t or i in used_d:
                continue
            used_t.add(tid)
            used_d.add(i)
            unmatched.discard(i)
            self._match(tid, detections[i], t)

        for i in unmatched:
            self._new(detections[i], t)

        for tid in ids:
            if tid not in used_t:
                tr = self.tracks[tid]
                tr.missed += 1
                if tr.missed > self.max_missed:
                    del self.tracks[tid]

        return [tr for tr in self.tracks.values() if tr.missed == 0]

    def _new(self, det, t):
        box, cls, score = det
        tid = self._next_id
        self._next_id += 1
        self.tracks[tid] = Track(tid, box, cls, score, t)

    def _match(self, tid, det, t):
        box, cls, score = det
        tr = self.tracks[tid]
        c = _centroid(box)
        dt = t - tr.last_time
        if dt > 0:
            px = _dist(tr.centroid, c)
            inst_kmh = (px * self.meters_per_pixel / dt) * 3.6
            inst_kmh = min(inst_kmh, 400.0)
            tr.speed_kmh = (tr.speed_kmh * (1 - self.smoothing) + inst_kmh * self.smoothing
                            if tr.speed_kmh else inst_kmh)
            tr.max_speed = max(tr.max_speed, tr.speed_kmh)
        tr.centroid = c
        tr.box = box
        tr.cls = cls
        tr.score = score
        tr.last_time = t
        tr.missed = 0
        tr.hits += 1

    def reset(self):
        self.tracks.clear()

File: web/src/lib/tracker.js
// Lightweight centroid tracker with speed estimation.
//
// Associates detections across frames using nearest-centroid matching with a
// distance gate, assigns stable IDs, and estimates speed from displacement of
// the centroid between timestamps using a pixel->meter calibration.

let NEXT_ID = 1;

function centroid(box) {
  return { x: box.x + box.w / 2, y: box.y + box.h / 2 };
}

function dist(a, b) {
  return Math.hypot(a.x - b.x, a.y - b.y);
}

export class Tracker {
  /**
   * @param {object} opts
   * @param {number} opts.maxDistance  max centroid distance (px) to match
   * @param {number} opts.maxMissed    frames a track survives without a match
   * @param {number} opts.metersPerPixel  calibration: real meters per pixel
   * @param {number} opts.smoothing    EMA factor for speed (0..1)
   */
  constructor(opts = {}) {
    this.maxDistance = opts.maxDistance ?? 120;
    this.maxMissed = opts.maxMissed ?? 12;
    this.metersPerPixel = opts.metersPerPixel ?? 0.05;
    this.smoothing = opts.smoothing ?? 0.4;
    this.tracks = new Map(); // id -> track
  }

  setCalibration(metersPerPixel) {
    this.metersPerPixel = metersPerPixel;
  }

  /**
   * @param {Array<{box,class,score}>} detections
   * @param {number} time timestamp in ms (e.g. performance.now or video.currentTime*1000)
   * @returns {Array} active tracks with id, box, class, score, speedKmh, color
   */
  update(detections, time) {
    const unmatched = new Set(detections.map((_, i) => i));
    const ids = [...this.tracks.keys()];

    // Greedy nearest matching: for each existing track, find closest detection.
    const pairs = [];
    for (const id of ids) {
      const t = this.tracks.get(id);
      let bestI = -1, bestD = Infinity;
      for (const i of unmatched) {
        const d = dist(t.centroid, centroid(detections[i].box));
        if (d < bestD) { bestD = d; bestI = i; }
      }
      if (bestI !== -1 && bestD <= this.maxDistance) {
        pairs.push({ id, i: bestI, d: bestD });
      }
    }
    pairs.sort((a, b) => a.d - b.d);

    const usedDet = new Set();
    const usedTrack = new Set();
    for (const p of pairs) {
      if (usedDet.has(p.i) || usedTrack.has(p.id)) continue;
      usedDet.add(p.i);
      usedTrack.add(p.id);
      unmatched.delete(p.i);
      this._matchTrack(p.id, detections[p.i], time);
    }

    // New tracks for unmatched detections.
    for (const i of unmatched) {
      this._newTrack(detections[i], time);
    }

    // Age out tracks that were not matched this frame.
    for (const id of ids) {
      if (!usedTrack.has(id)) {
        const t = this.tracks.get(id);
        t.missed += 1;
        if (t.missed > this.maxMissed) this.tracks.delete(id);
      }
    }

    return [...this.tracks.values()]
      .filter((t) => t.missed === 0)
      .map((t) => ({
        id: t.id,
        box: t.box,
        class: t.class,
        score: t.score,
        speedKmh: t.speedKmh,
        color: t.color,
        plate: t.plate,
        plateBox: t.plateBox,
        hits: t.hits,
        thumb: t.thumb,
        plateThumb: t.plateThumb,
      }));
  }

  _newTrack(det, time) {
    const id = NEXT_ID++;
    this.tracks.set(id, {
      id,
      box: det.box,
      class: det.class,
      score: det.score,
      centroid: centroid(det.box),
      lastTime: time,
      speedKmh: 0,
      missed: 0,
      color: null,
      plate: null,
      plateBox: null,
      maxSpeed: 0,
      hits: 1,
    });
  }

  _matchTrack(id, det, time) {
    const t = this.tracks.get(id);
    const c = centroid(det.box);
    const dt = (time - t.lastTime) / 1000; // seconds
    if (dt > 0) {
      const pxMoved = dist(t.centroid, c);
      const metersMoved = pxMoved * this.metersPerPixel;
      const instMs = metersMoved / dt; // m/s
      const instKmh = instMs * 3.6;
      // EMA smoothing; clamp absurd values.
      const clamped = Math.min(instKmh, 400);
      t.speedKmh = t.speedKmh
        ? t.speedKmh * (1 - this.smoothing) + clamped * this.smoothing
        : clamped;
      t.maxSpeed = Math.max(t.maxSpeed, t.speedKmh);
    }
    t.centroid = c;
    t.box = det.box;
    t.class = det.class;
    t.score = det.score;
    t.lastTime = time;
    t.missed = 0;
    t.hits += 1;
  }

  // Attach metadata produced by async stages (color, plate OCR).
  setMeta(id, meta) {
    const t = this.tracks.get(id);
    if (t) Object.assign(t, meta);
  }

  reset() {
    this.tracks.clear();
  }
}
